Approximate convex decomposition

soesau opened the pull request:
## Summary of Changes

Reimplementation of https://github.com/kmammou/v-hacd
Performs approximate convex decomposition of an input mesh based on a voxelization and hierarchical decomposition.

~~Although it could have added to the package  _Convex Decomposition of Polyhedra_   it was decided to add it to the package _Polygon Mesh Processing_.~~
Finally we decided that it gets added to   _Convex Decomposition of Polyhedra_  

## Release Management

* Affected package(s): PMP
* Feature/Small Feature (if any): [Approximate_Convex_Decomposition](https://cgalwiki.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Approximate_Convex_Decomposition)
* Link to compiled documentation: [v0](https://cgal.github.io/8935/v0/Polygon_mesh_processing/group__PMP__convex__decomposition__grp.html)

## TODO
- [x] cite: Khaled Mamou. “Volumetric Hierarchical Approximate Convex Decomposition”. In Game
Engine Gems 3, A K Peters, 2016, pp. 141–158. (It would be usefull if you find a pdf of it)
- [x] <s>Eventually cite: https://arxiv.org/pdf/2205.02961 (optional, it's a possible improvement)</s>
- [x] Produce a plugin for the demo
- [x] Produce a postprocess to refine the convex hull to avoid an offset compare to the original mesh (optional)

soesau committed:
first commit of working approximate convex decomposition

soesau committed:
clean up and moving results into outputiterator

soesau committed:
updated dependencies
Surface_mesh will be removed later on

soesau commented:
The dependency to Surface_mesh will be removed later.

soesau committed:
fixed bugs in the splitting method
added rayshooting to fill inside and outside of unclosed meshes
tweaking the find best splitting plane method

soesau committed:
parallelization of ray shooting and hierarchical recursion

soesau committed:
moving saving into example

soesau committed:
clean up

soesau committed:
cleanup
adding documentation

soesau committed:
Merge remote-tracking branch 'cgal/master' into PMP-approximate_convex_decomposition-GF

afabri commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/approximate_convex_decomposition.h:
```diff
@@ -1394,9 +1364,82 @@ void merge(std::vector<Convex_hull<GeomTraits>>& candidates, std::vector<int8_t>
+ * \ingroup PMP_convex_decomposition_grp
+ *
+ * \brief approximates the input mesh by a number of convex hulls. The input mesh is voxelized and the convex hull of the mesh is hierarchically split until a volume threshold is met.
+ *        Afterwards, a greedy pair-wise merging combines smaller convex hulls until the given number of convex hulls is met.
```
```suggestion
 *        Afterwards, a greedy pairwise merging combines smaller convex hulls until the given number of convex hulls is met.
```

afabri committed:
typo

afabri commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/approximate_convex_decomposition.h:
```diff
@@ -1394,9 +1364,82 @@ void merge(std::vector<Convex_hull<GeomTraits>>& candidates, std::vector<int8_t>
+ *   \cgalParamNEnd
+ *
+ *   \cgalParamNBegin{split_at_concavity}
+ *     \cgalParamDescription{split the local box of the convex hull in the mid of the longest axis (faster) or search the concavity along the largest axis for splitting.}
```
Is longest and largest the same?  Then use only one word.

soesau replied:
fixed

soesau committed:
cleaning code and doc

soesau committed:
adding test
making compatible with EPECK

soesau committed:
fixing warnings

soesau committed:
updated CHANGES.md

soesau committed:
typo fix

soesau committed:
Merge branch 'master' into PMP-approximate_convex_decomposition-GF

soesau committed:
pass on doc

sloriot commented:
Warnings in PMP tests/exs in CGAL-6.1-Ic-191

soesau committed:
fixed warnings
reduced test complexity (to prevent timeout)

soesau committed:
fix test

soesau committed:
fix of detection of TBB
shortened EPECK test

soesau committed:
removing unused variable

soesau committed:
Merge branch 'master' into PMP-approximate_convex_decomposition-GF

afabri commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/approximate_convex_decomposition.h:
```diff
@@ -0,0 +1,1689 @@
+#ifndef CGAL_POLYGON_MESH_PROCESSING_CONVEX_DECOMPOSITION_H
+#define CGAL_POLYGON_MESH_PROCESSING_CONVEX_DECOMPOSITION_H
+
+#include <CGAL/license/Polygon_mesh_processing/combinatorial_repair.h>
```
@sloriot do you approve that is goes in this subpackage of PMP ?

afabri commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/approximate_convex_decomposition.h:
```diff
@@ -0,0 +1,1689 @@
+template<typename GeomTraits>
+Iso_cuboid_3<GeomTraits> bbox_voxel(const Vec3_uint& voxel, const Bbox_3& bb, const typename GeomTraits::FT& voxel_size) {
+  return Bbox_3(
+    bb.xmin() + voxel[0] * to_double(voxel_size),
```
Wouldn't it make sense to call `to_double()` only once, even outside of this function?

soesau replied:
Done, but inside of the function.

afabri commented on Polygon_mesh_processing/examples/Polygon_mesh_processing/approximate_convex_decomposition.cpp:
```diff
@@ -0,0 +1,46 @@
+    return 1;
+  }
+
+  std::vector<Convex_hull> convex_hulls;
```
Maybe `reserve(9)`

afabri commented on Installation/CHANGES.md:
```diff
@@ -1,9 +1,15 @@
+## [Release 6.2](https://github.com/CGAL/cgal/releases/tag/v6.2)
+
+### [Polygon Mesh Processing](https://doc.cgal.org/6.2/Manual/packages.html#PkgPolygonMeshProcessing)
+- Added the function `CGAL::Polygon_mesh_processing::approximate_convex_decomposition` that decomposes an input mesh into a specified number of convex hulls. It performs an hierarchical splitting of the convex hull guided by an volumetric error and performs a subsequent merging of convex hulls to obtain the targeted number.
```
```suggestion
- Added the function `CGAL::Polygon_mesh_processing::approximate_convex_decomposition()` that decomposes an input mesh into a specified number of convex hulls. It performs a hierarchical splitting of the convex hull guided by an volumetric error and performs a subsequent merging of convex hulls to obtain the targeted number.
```

afabri commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/approximate_convex_decomposition.h:
```diff
@@ -0,0 +1,1689 @@
+
+  stream <<
+    "ply" << std::endl <<
+    "format ascii 1.0" << std::endl <<
```
Maybe replace all `std::endl` with a `\n`  to avoid flushing.

afabri commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/approximate_convex_decomposition.h:
```diff
@@ -0,0 +1,1689 @@
+  Bbox_uint(const Vec3_uint &lower, const Vec3_uint &upper) : lower(lower), upper(upper) {}
+};
+
+enum Grid_cell : int8_t {
```
We could as well use `CGAL::Bounded_side`.   Is it important that it is `int_8` because this is the type of a data member?

afabri commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/approximate_convex_decomposition.h:
```diff
@@ -0,0 +1,1689 @@
+}
+
+// Valid voxel grids separate OUTSIDE from INSIDE via SURFACE
+bool check_grid(std::vector<int8_t>& grid, const Vec3_uint& grid_size) {
```
Is this an `is_valid()` function? Then rename it.

afabri reviewed (changes requested):
Sorry about the non-batched review comments.

Please add comment such as `} // namespace internal`

I am wondering if class names like `Set` are not too generic.
Also I was surprised to see a `struct Convex_hull`.

soesau commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/approximate_convex_decomposition.h:
```diff
@@ -0,0 +1,1689 @@
+
+  FT s = 1.5 * voxel_size;
+
+  bb = Bbox_3(to_double(bb.xmin() - s), to_double(bb.ymin() - s), to_double(bb.zmin() - s), to_double(bb.xmax() + s), to_double(bb.ymax() + s), to_double(bb.zmax() + s));
```
Only if the bounding box is a cube.

soesau committed:
integrate reviews from comments

afabri committed:
Update STL_Extension/include/CGAL/STL_Extension/internal/parameters_interface.h

soesau committed:
integrating reviews

sloriot commented:
Error [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.1-Ic-221/Polygon_mesh_processing/TestReport_magritte_arm64-apple_bigsur_clang1200-release-64bits.gz)

sloriot commented:
This is the output I get for the elephant. Is it expected?
<img width="877" height="853" alt="image" src="https://github.com/user-attachments/assets/229213fe-2e94-4211-976d-789400233ba6" />

soesau committed:
bugfixes

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/approximate_convex_decomposition.h:
```diff
@@ -0,0 +1,1672 @@
+#include <CGAL/Polygon_mesh_processing/measure.h>
+#include <CGAL/utils_classes.h>
+
+#include <CGAL/Surface_mesh.h>
```
Then you should remove it, no?

soesau committed:
Merge remote-tracking branch 'cgal/master' into PMP-approximate_convex_decomposition-GF

soesau committed:
switching from convex_hull_3 using Surface_mesh to indexed triangle list
removing Surface_mesh dependency

soesau committed:
removing use of Surface_mesh

MaelRL commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/approximate_convex_decomposition.h:
```diff
@@ -0,0 +1,1639 @@
+ *
+ *   \cgalParamNBegin{maximum_number_of_voxels}
+ *     \cgalParamDescription{gives an upper bound of the number of voxels. The longest bounding box side will have a length of the cubic root of `maximum_number_of_voxels` rounded down. Cannot be smaller than 64. }
+ *     \cgalParamType{unsigned int}
```
you use std::size_t in the code

MaelRL commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/approximate_convex_decomposition.h:
```diff
@@ -0,0 +1,1639 @@
+ *
+ * \cgalNamedParamsEnd
+ *
+ * \return number of convex hulls. Can be lower than the `maximum_number_of_convex_hulls` if the specified `volume_error` is quickly met.
```
```suggestion
 * \return the number of convex hulls. Note that this value may be lower than the `maximum_number_of_convex_hulls`, for example if the specified `volume_error` is quickly met.
```

MaelRL commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/approximate_convex_decomposition.h:
```diff
@@ -0,0 +1,1639 @@
+ * \ingroup PMP_convex_decomposition_grp
+ *
+ * \brief approximates the input mesh by a number of convex hulls. The input mesh is voxelized and the voxels intersecting with the mesh are labeled as surface.
+ *        The remaining voxels are labeled as outside or inside by floodfill, in case the input mesh is closed, or by axis-aligned ray shooting, i.e., along x,
```
```suggestion
 *        The remaining voxels are labeled as outside or inside by flood fill, in case the input mesh is closed, or by axis-aligned ray shooting, i.e., along x,
```

MaelRL commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/approximate_convex_decomposition.h:
```diff
@@ -0,0 +1,1639 @@
+ *   \cgalParamNBegin{concurrency_tag}
+ *     \cgalParamDescription{a tag indicating if the task should be done using one or several threads.}
+ *     \cgalParamType{Either `CGAL::Sequential_tag`, or `CGAL::Parallel_tag`, or `CGAL::Parallel_if_available_tag`}
+ *     \cgalParamDefault{`CGAL::Parallel_if_available_tag`}
```
So far, `concurrency_tag`, whether it was passed as a template parameter or NP, has always defaulted to `Sequential_tag` in the header (and `Parallel_if_available_tag` can be in the example), with IIRC the reasoning being that it is simpler, for example to control exactly at which level you want to parallelize a stack of potentially parallel functions.

soesau committed:
integrating review
cleaning up

soesau committed:
Apply suggestions from code review

Co-authored-by: Mael <mael.rouxel.labbe@geometryfactory.com>

soesau committed:
added demo plugin for approximate convex decomposition

soesau committed:
bugfix removed CGAL_USE

soesau committed:
created Surface_mesh_decomposition package
added approximate_convex_decomposition doc to Convex_decomposition_3
added license file

soesau committed:
Merge branch 'main' into PMP-approximate_convex_decomposition-GF

soesau committed:
added reference

afabri commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/approximate_convex_decomposition.h:
```diff
@@ -0,0 +1,1622 @@
+  using Geom_traits = typename GetGeomTraits<FaceGraph, NamedParameters>::type;
+
+  using FT = typename Geom_traits::FT;
+  const unsigned int num_voxels = parameters::choose_parameter(parameters::get_parameter(np, internal_np::maximum_number_of_voxels), 1000000);
```
Is that a secret parameter?

soesau committed:
fixed cmake project name

soesau commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/approximate_convex_decomposition.h:
```diff
@@ -0,0 +1,1622 @@
+  Candidate() : depth(0), bbox({ 0, 0, 0 }, { 0, 0, 0 }) {index = cidx++;}
+  Candidate(std::size_t depth, Bbox_uint &bbox) : depth(depth), bbox(bbox) { index = cidx++; }
+private:
+  inline static std::size_t cidx = 0;
```
No, but no new instances are created in parallel.

soesau committed:
fixing named parameter
removing approximate convex decomposition from PMP

soesau committed:
adapting dependencies

soesau committed:
fixed include

soesau committed:
fixed namespace

soesau committed:
removed unused variable

soesau commented:
/build:v0

soesau committed:
removed old references to approximate convex decomposition

soesau committed:
doc fix

soesau committed:
remove 'we' and 'our'
added implementation history

soesau committed:
doc update
bugfix

soesau committed:
adding text to examples

soesau committed:
corrected numbers in table

soesau committed:
adapted named parameter doc
[skip ci]

soesau commented:
/build:v0

afabri commented on Surface_mesh_decomposition/include/CGAL/approximate_convex_decomposition.h:
```diff
@@ -706,7 +706,7 @@ void rayshooting_fill(std::vector<int8_t>& grid, const Vec3_uint& grid_size, con
               face_descriptor fd = intersection->second;
-              Vector_3 n = compute_face_normal(fd, mesh);
+              Vector_3 n = Polygon_mesh_processing::compute_face_normal(fd, mesh);
               if (dirs[i] * n > 0)
```
[`Angle_3`](https://doc.cgal.org/latest/Kernel_23/classKernel_1_1Angle__3.html)

sloriot replied:
@soesau do I get it right that here you want to classify points from a grid? THat is whether they are inside or outside of a domain bounded by a mesh?

soesau committed:
moving example from Surface_mesh_decomposition into Convex_decomposition_3

soesau commented:
/build:v0

afabri committed:
Improve documentation of Nef

afabri committed:
Improve documentation of Nef

afabri committed:
Improve documentation of Nef

soesau committed:
added refitting

afabri commented:
I had thought that the vertices of the  decomposition are all input vertices. Is the current solution (intersection points between voxels and triangles) also what is done in the original paper?

soesau commented:
I use the AABB_tree for collecting all intersections inside an Iso_cuboid_3. I do some simple checks for which corners of the voxel grid I also have to include for the ch3.

soesau committed:
making refitting more robust for coarse voxel grids

soesau committed:
integrating review
removing debugging leftover

soesau committed:
doc and benchmark update

soesau committed:
adding result image
changing overview image
adapting test

soesau committed:
adding images

soesau committed:
typos
[skip ci]

soesau committed:
added note that output may contain degenerated volumes
[skip ci]

soesau committed:
doc update

afabri commented on Installation/CHANGES.md:
```diff
@@ -11,6 +11,10 @@ Release date: July 2026
 
+### [Convex Decomposition of Polyhedra](https://doc.cgal.org/6.2/Manual/packages.html#PkgConvexDecomposition3)
+- Added the function `CGAL::approximate_convex_decomposition()` that computes a set of convex volumes that cover an input mesh.
+  It performs a hierarchical splitting of the convex hull guided by an volumetric error and performs a subsequent merging of convex hulls to obtain the targeted count.
```
```suggestion
```
I suggest to remove the description how it works, as the change log just states what gets added.  @sloriot do you agree?

afabri reviewed (approved):
I approve this small feature, with two changes to be made, 
There should be no link to functions in the package description, and the change log should be shortened.

soesau committed:
doc changes
[skip ci]

soesau committed:
removed test in old location (new test is in Surface_mesh_decomposition)

soesau committed:
fix warnings

albert-github commented on Convex_decomposition_3/doc/Convex_decomposition_3/Doxyfile.in:
```diff
@@ -3,3 +3,5 @@
 EXTRACT_ALL                =  false
 HIDE_UNDOC_CLASSES         =  true
+
+INPUT += ${CGAL_Surface_mesh_decomposition_INCLUDE_DIR}/CGAL/approximate_convex_decomposition.h
```
Overnight documentation build gives:
```
warning: source '/CGAL/approximate_convex_decomposition.h' is not a readable file or directory... skipping.
```

Maybe `${CGAL_PACKAGE_INCLUDE_DIR}` or the file is not in the indicated place / variable `CGAL_Surface_mesh_decomposition_INCLUDE_DIR` does not exist.

afabri committed:
Apply suggestion from @afabri

sloriot committed:
fix doc include in non branch build mode

sloriot committed:
whitespaces

afabri committed:
fix fixture

sloriot committed:
missing inline

sloriot committed:
more missing inline
